File: matrix.py
# matrix.py
import json
from dataclasses import dataclass, asdict
from enum import Enum
from typing import List, Optional
from datetime import datetime
from pathlib import Path

class Quadrant(Enum):
    DO = "Важно и срочно"
    PLAN = "Важно, но не срочно"
    DELEGATE = "Не важно, но срочно"
    ELIMINATE = "Не важно и не срочно"

@dataclass
class Task:
    id: int
    title: str
    description: str
    important: bool
    urgent: bool
    done: bool
    created_at: str

    def quadrant(self) -> Quadrant:
        if self.important and self.urgent:
            return Quadrant.DO
        elif self.important and not self.urgent:
            return Quadrant.PLAN
        elif not self.important and self.urgent:
            return Quadrant.DELEGATE
        else:
            return Quadrant.ELIMINATE

class EisenhowerMatrix:
    def __init__(self):
        self.tasks: List[Task] = []
        self.next_id = 1

    def add_task(self, title: str, desc: str, important: bool, urgent: bool) -> Task:
        task = Task(
            id=self.next_id,
            title=title,
            description=desc,
            important=important,
            urgent=urgent,
            done=False,
            created_at=datetime.now().isoformat()
        )
        self.tasks.append(task)
        self.next_id += 1
        return task

    def find_task(self, task_id: int) -> Optional[Task]:
        return next((t for t in self.tasks if t.id == task_id), None)

    def edit_task(self, task_id: int, title: str = None, desc: str = None,
                  important: bool = None, urgent: bool = None) -> bool:
        task = self.find_task(task_id)
        if not task:
            return False
        if title is not None:
            task.title = title
        if desc is not None:
            task.description = desc
        if important is not None:
            task.important = important
        if urgent is not None:
            task.urgent = urgent
        return True

    def toggle_done(self, task_id: int) -> bool:
        task = self.find_task(task_id)
        if not task:
            return False
        task.done = not task.done
        return True

    def delete_task(self, task_id: int) -> bool:
        task = self.find_task(task_id)
        if task:
            self.tasks.remove(task)
            return True
        return False

    def filter_by_quadrant(self, quadrant: Quadrant) -> List[Task]:
        return [t for t in self.tasks if t.quadrant() == quadrant]

    def filter_by_done(self, done: bool) -> List[Task]:
        return [t for t in self.tasks if t.done == done]

    def get_stats(self) -> dict:
        total = len(self.tasks)
        done_count = sum(1 for t in self.tasks if t.done)
        quad_counts = {q: 0 for q in Quadrant}
        for t in self.tasks:
            quad_counts[t.quadrant()] += 1
        return {
            "total": total,
            "done": done_count,
            "pending": total - done_count,
            "quadrants": quad_counts
        }

    def save_to_file(self, filename: str = "eisenhower_data.json") -> None:
        with open(filename, "w", encoding="utf-8") as f:
            json.dump([asdict(t) for t in self.tasks], f, ensure_ascii=False, indent=2)

    def load_from_file(self, filename: str = "eisenhower_data.json") -> None:
        path = Path(filename)
        if not path.exists():
            return
        with open(filename, "r", encoding="utf-8") as f:
            data = json.load(f)
            self.tasks.clear()
            for item in data:
                task = Task(
                    id=item["id"],
                    title=item["title"],
                    description=item["description"],
                    important=item["important"],
                    urgent=item["urgent"],
                    done=item["done"],
                    created_at=item["created_at"]
                )
                self.tasks.append(task)
                if task.id >= self.next_id:
                    self.next_id = task.id + 1

def print_task(task: Task, show_quadrant: bool = True) -> None:
    done_mark = "✅" if task.done else "⬜"
    print(f"{done_mark} #{task.id} - {task.title}")
    print(f"   Описание: {task.description}")
    if show_quadrant:
        q = task.quadrant()
        print(f"   Квадрант: {q.value} (важно: {task.important}, срочно: {task.urgent})")
    print(f"   Создана: {task.created_at}")

def main():
    matrix = EisenhowerMatrix()
    matrix.load_from_file()

    while True:
        print("\n===== МАТРИЦА ЭЙЗЕНХАУЭРА (Python) =====")
        print("1. Показать все задачи")
        print("2. Добавить задачу")
        print("3. Редактировать задачу")
        print("4. Отметить выполненной / невыполненной")
        print("5. Удалить задачу")
        print("6. Показать статистику")
        print("7. Сохранить в файл")
        print("8. Загрузить из файла")
        print("0. Выход")
        choice = input("Выберите действие: ").strip()

        if choice == "0":
            break
        elif choice == "1":
            if not matrix.tasks:
                print("Задач нет.")
                continue
            # Показать с группировкой по квадрантам
            for q in Quadrant:
                tasks = matrix.filter_by_quadrant(q)
                if tasks:
                    print(f"\n--- {q.value} ---")
                    for t in tasks:
                        print_task(t, show_quadrant=False)
        elif choice == "2":
            title = input("Введите название: ").strip()
            if not title:
                print("Название не может быть пустым.")
                continue
            desc = input("Введите описание: ").strip()
            important = input("Задача важна? (y/n): ").strip().lower() == 'y'
            urgent = input("Задача срочна? (y/n): ").strip().lower() == 'y'
            task = matrix.add_task(title, desc, important, urgent)
            print(f"Задача добавлена с ID {task.id}, квадрант: {task.quadrant().value}")
        elif choice == "3":
            try:
                task_id = int(input("Введите ID задачи: ").strip())
            except ValueError:
                print("Некорректный ID.")
                continue
            task = matrix.find_task(task_id)
            if not task:
                print("Задача не найдена.")
                continue
            print("Оставьте поле пустым, чтобы не менять.")
            new_title = input(f"Новое название ({task.title}): ").strip()
            if new_title:
                task.title = new_title
            new_desc = input(f"Новое описание ({task.description}): ").strip()
            if new_desc:
                task.description = new_desc
            new_imp = input(f"Важна? (y/n) сейчас: {'y' if task.important else 'n'}): ").strip()
            if new_imp:
                task.important = new_imp.lower() == 'y'
            new_urg = input(f"Срочна? (y/n) сейчас: {'y' if task.urgent else 'n'}): ").strip()
            if new_urg:
                task.urgent = new_urg.lower() == 'y'
            print("Задача обновлена.")
        elif choice == "4":
            try:
                task_id = int(input("Введите ID задачи: ").strip())
            except ValueError:
                print("Некорректный ID.")
                continue
            if matrix.toggle_done(task_id):
                print("Статус выполнения изменён.")
            else:
                print("Задача не найдена.")
        elif choice == "5":
            try:
                task_id = int(input("Введите ID задачи для удаления: ").strip())
            except ValueError:
                print("Некорректный ID.")
                continue
            if matrix.delete_task(task_id):
                print("Задача удалена.")
            else:
                print("Задача не найдена.")
        elif choice == "6":
            stats = matrix.get_stats()
            print("\n=== СТАТИСТИКА ===")
            print(f"Всего задач: {stats['total']}")
            print(f"Выполнено: {stats['done']}")
            print(f"Осталось: {stats['pending']}")
            for q, count in stats['quadrants'].items():
                print(f"  {q.value}: {count}")
        elif choice == "7":
            matrix.save_to_file()
            print("Сохранено.")
        elif choice == "8":
            matrix.load_from_file()
            print("Загружено.")
        else:
            print("Неизвестная команда.")

if __name__ == "__main__":
    main()

File: matrix.rb
# matrix.rb
require 'json'
require 'date'

class Quadrant
  DO = "Важно и срочно"
  PLAN = "Важно, но не срочно"
  DELEGATE = "Не важно, но срочно"
  ELIMINATE = "Не важно и не срочно"
end

class Task
  attr_accessor :id, :title, :description, :important, :urgent, :done, :created_at

  def initialize(id, title, description, important, urgent, done = false, created_at = DateTime.now.iso8601)
    @id = id
    @title = title
    @description = description
    @important = important
    @urgent = urgent
    @done = done
    @created_at = created_at
  end

  def quadrant
    if @important && @urgent
      Quadrant::DO
    elsif @important && !@urgent
      Quadrant::PLAN
    elsif !@important && @urgent
      Quadrant::DELEGATE
    else
      Quadrant::ELIMINATE
    end
  end

  def to_h
    {
      id: @id,
      title: @title,
      description: @description,
      important: @important,
      urgent: @urgent,
      done: @done,
      created_at: @created_at
    }
  end

  def self.from_h(hash)
    Task.new(hash[:id], hash[:title], hash[:description], hash[:important], hash[:urgent], hash[:done], hash[:created_at])
  end
end

class EisenhowerMatrix
  attr_reader :tasks

  def initialize
    @tasks = []
    @next_id = 1
  end

  def add_task(title, description, important, urgent)
    task = Task.new(@next_id, title, description, important, urgent)
    @tasks << task
    @next_id += 1
    task
  end

  def find_task(id)
    @tasks.find { |t| t.id == id }
  end

  def edit_task(id, title: nil, description: nil, important: nil, urgent: nil)
    task = find_task(id)
    return false unless task
    task.title = title if title
    task.description = description if description
    task.important = important unless important.nil?
    task.urgent = urgent unless urgent.nil?
    true
  end

  def toggle_done(id)
    task = find_task(id)
    return false unless task
    task.done = !task.done
    true
  end

  def delete_task(id)
    task = find_task(id)
    return false unless task
    @tasks.delete(task)
    true
  end

  def filter_by_quadrant(quadrant)
    @tasks.select { |t| t.quadrant == quadrant }
  end

  def filter_by_done(done)
    @tasks.select { |t| t.done == done }
  end

  def stats
    total = @tasks.size
    done_count = @tasks.count(&:done)
    quad_counts = Hash.new(0)
    @tasks.each { |t| quad_counts[t.quadrant] += 1 }
    {
      total: total,
      done: done_count,
      pending: total - done_count,
      quadrants: quad_counts
    }
  end

  def save_to_file(filename = "eisenhower_data.json")
    File.write(filename, JSON.pretty_generate(@tasks.map(&:to_h)))
  end

  def load_from_file(filename = "eisenhower_data.json")
    return unless File.exist?(filename)
    data = JSON.parse(File.read(filename), symbolize_names: true)
    @tasks.clear
    data.each do |item|
      task = Task.from_h(item)
      @tasks << task
      @next_id = task.id + 1 if task.id >= @next_id
    end
  rescue JSON::ParserError
    puts "Ошибка чтения файла."
  end
end

def print_task(task, show_quadrant = true)
  done_mark = task.done ? "✅" : "⬜"
  puts "#{done_mark} ##{task.id} - #{task.title}"
  puts "   Описание: #{task.description}"
  if show_quadrant
    puts "   Квадрант: #{task.quadrant} (важно: #{task.important}, срочно: #{task.urgent})"
  end
  puts "   Создана: #{task.created_at}"
end

def main
  matrix = EisenhowerMatrix.new
  matrix.load_from_file

  loop do
    puts "\n===== МАТРИЦА ЭЙЗЕНХАУЭРА (Ruby) ====="
    puts "1. Показать все задачи"
    puts "2. Добавить задачу"
    puts "3. Редактировать задачу"
    puts "4. Отметить выполненной / невыполненной"
    puts "5. Удалить задачу"
    puts "6. Показать статистику"
    puts "7. Сохранить в файл"
    puts "8. Загрузить из файла"
    puts "0. Выход"
    print "Выберите действие: "
    choice = gets.chomp

    case choice
    when "0"
      break
    when "1"
      if matrix.tasks.empty?
        puts "Задач нет."
      else
        [Quadrant::DO, Quadrant::PLAN, Quadrant::DELEGATE, Quadrant::ELIMINATE].each do |q|
          tasks = matrix.filter_by_quadrant(q)
          if tasks.any?
            puts "\n--- #{q} ---"
            tasks.each { |t| print_task(t, false) }
          end
        end
      end
    when "2"
      print "Введите название: "
      title = gets.chomp
      next if title.empty?
      print "Введите описание: "
      desc = gets.chomp
      print "Задача важна? (y/n): "
      important = gets.chomp.downcase == 'y'
      print "Задача срочна? (y/n): "
      urgent = gets.chomp.downcase == 'y'
      task = matrix.add_task(title, desc, important, urgent)
      puts "Задача добавлена с ID #{task.id}, квадрант: #{task.quadrant}"
    when "3"
      print "Введите ID задачи: "
      id = gets.chomp.to_i
      task = matrix.find_task(id)
      unless task
        puts "Задача не найдена."
        next
      end
      puts "Оставьте поле пустым, чтобы не менять."
      print "Новое название (#{task.title}): "
      new_title = gets.chomp
      task.title = new_title unless new_title.empty?
      print "Новое описание (#{task.description}): "
      new_desc = gets.chomp
      task.description = new_desc unless new_desc.empty?
      print "Важна? (y/n) сейчас: #{task.important ? 'y' : 'n'}: "
      new_imp = gets.chomp
      task.important = (new_imp.downcase == 'y') unless new_imp.empty?
      print "Срочна? (y/n) сейчас: #{task.urgent ? 'y' : 'n'}: "
      new_urg = gets.chomp
      task.urgent = (new_urg.downcase == 'y') unless new_urg.empty?
      puts "Задача обновлена."
    when "4"
      print "Введите ID задачи: "
      id = gets.chomp.to_i
      if matrix.toggle_done(id)
        puts "Статус выполнения изменён."
      else
        puts "Задача не найдена."
      end
    when "5"
      print "Введите ID задачи для удаления: "
      id = gets.chomp.to_i
      if matrix.delete_task(id)
        puts "Задача удалена."
      else
        puts "Задача не найдена."
      end
    when "6"
      stats = matrix.stats
      puts "\n=== СТАТИСТИКА ==="
      puts "Всего задач: #{stats[:total]}"
      puts "Выполнено: #{stats[:done]}"
      puts "Осталось: #{stats[:pending]}"
      stats[:quadrants].each do |q, count|
        puts "  #{q}: #{count}"
      end
    when "7"
      matrix.save_to_file
      puts "Сохранено."
    when "8"
      matrix.load_from_file
      puts "Загружено."
    else
      puts "Неизвестная команда."
    end
  end
end

main if __FILE__ == $0
